### DISCUSSION TITLE
Percent deviance explained

### DISCUSSION THREAD

**Discussion by James-Thorson-NOAA** [sdmTMB/sdmTMB#331] [Ideas]
@seananderson I was thinking about adding `deviance_explained` as function to tinyVAST.  Similar to VAST, it would use the formula for relative deviance conditional on ML and EB estimators.  I have the relative-deviance formula for most but not all likelihoods, and it matches output from mgcv for those cases (where mgcv typically uses plug-in estimators, e.g., conditional calculations).  Is it worth discussing how to add a similar function to `sdmTMB`?

**Comment by seananderson** [comment]
Yes, I'd be very interested in getting something like this in the package.

**Comment by James-Thorson-NOAA** [comment]
In case it's helpful, I went ahead and prototyped it in tinyVAST@dev here: https://github.com/vast-lib/tinyVAST/blob/dev/R/utility.R$L333

I have several checks that confirm that it matches glm / mgcv in the vignettes and CI checks.

**Comment by James-Thorson-NOAA** [comment]
Do you wanna look at the tinyVAST demo and respond with some thoughts on interface, or would you rather I dive in and propose something via branch and PR?  

I'd of course prefer the former, just to avoid the extra work in case you have improvements upon the tinyVAST implementation that you'd catch early in the process :)

**Comment by slamander** [comment]
Hi @James-Thorson-NOAA & @seananderson! 
Have there been any updates in this area? I'm trying to calculate conditional percent deviance explained for sdmTMB, but I'm curious what route would be most appropriate. I can't use the typical method for calculating "conditional" deviance (i.e., comparing to a saturated model; `stats::deviance()`), so I've instead calculated "unconditional" deviance (`-2*logLik(obj)`). Are either of these appropriate for use with spatial models with dispersion parameters?

**Comment by James-Thorson-NOAA** [comment]
The "unconditional" deviance definitely isn't right.  I have the deviance-residuals for normal, lognormal, Poisson, tweedie, gamma, and negbin1 or negbin2 families in tinyVAST [here](https://github.com/vast-lib/tinyVAST/blob/main/src/tinyVAST.cpp#L522).

@seananderson any ideas how you'd want the conditional-deviance to be included (presumably in the CPP side) and accessed in terms of S3 generics or a new helper function?

**Comment by seananderson** [comment]
I think I've got this working for single linear predictor models based on your code in tinyVAST, @James-Thorson-NOAA. I'll need to finish the unit tests and adapt for delta models. Perhaps making a `deviance.sdmTMB()` S3 method would be the simplest place to start to match `glm()`. A user can then fit a null model and compare. After getting that right we could decide whether to write an equivalent to the tinyVAST `deviance_explained()` as a helper.

**Comment by James-Thorson-NOAA** [comment]
Awesome!  tinyVAST uses the sign(yhat - y) * sqrt(relative_deviance) as the deviance-residual ... those match lm/glm for the cases I looked, and also matches the definition I've seen online.  But I'm not sure how to define the deviance-residual for a delta model, so that equivalence between relative deviance and deviance residuals breaks down I think for delta models.  

But for delta models, you could compute the relative deviance for each sample as the sum of the relative deviance for the 0/1 and the relative deviance of the other component when nonzero.  I can't remember if I actually compared that calculation with the relative deviance from pscl (an alternative implementation for delta models), but it makes sense on a theory basis.

Do you want me to help with anything?  Sorry as always that I don't know how best to help with tag-teaming the coding side :0

**Comment by seananderson** [comment]
This should be in the main branch now. For now, you can calculate deviance explained yourself by fitting and comparing to a null model, but I'll consider a helper function or wrapping it into the default `summary()` output. Demo:

``` r
library(sdmTMB)
mesh <- make_mesh(pcod, c("X", "Y"), cutoff = 10)
fit <- sdmTMB(
  density ~ 1,
  spatial = "on", 
  mesh = mesh,
  family = delta_gamma(), 
  data = pcod,
)

# model with *only* intercept and no random fields:
fit_null <- sdmTMB(
  density ~ 1, 
  spatial = "off", 
  family = delta_gamma(),
  data = pcod,
)

# proportion deviance explained:
devexplained <- 1 - deviance(fit) / deviance(fit_null)
devexplained
#> [1] 0.3204026

# showing that tinyVAST and sdmTMB match on delta model deviance:
m_sdmTMB <- sdmTMB(
  density ~ depth_scaled, 
  family = delta_gamma(type = "poisson-link"), 
  spatial = "off", 
  data = pcod
)
deviance(m_sdmTMB)
#> [1] 4980.735

m_tinyVAST <- tinyVAST::tinyVAST(
  density ~ depth_scaled,
  family = tinyVAST::delta_gamma(type = "poisson-link"),
  data = as.data.frame(pcod), 
  delta_options = list(formula = ~depth_scaled)
)
r <- m_tinyVAST$obj$report()
r$deviance
#> [1] 4980.735
```

<sup>Created on 2025-05-12 with [reprex v2.1.1.9000](https://reprex.tidyverse.org)</sup>

**Comment by James-Thorson** [comment]
Thanks Sean for adding deviance residuals, allowing percent deviance explained!

FWIW, I just updated tinyVAST@dev [here](https://github.com/vast-lib/tinyVAST/blob/dev/src/tinyVAST.cpp#L530) to have the deviance-residual calculation for family = "binomial" with weights >= 2, which appears to be disabled in sdmTMB.  No need to add from my end, but thought I'd point out the source for future reference.

**Reply by seananderson** [reply]
Ah, thanks. I'll put it on my todo list to incorporate that. I was also looking into the generalized gamma deviance, but haven't tested anything yet.

**Comment by KGallagher7** [comment]
Hi @seananderson & @James-Thorson! A quick follow up question on this - could you use this method to calculate the deviance explained by different covariates in the models within sdmTMB (i.e. make a model with just a single variable and calculate the deviance explained from that model)? I've used this method for GAMMs in the past, but not sure if it could also be used for the individual components in sdmTMB/tinyVAST. Thanks!

**Comment by James-Thorson-NOAA** [comment]
There's many ways to quantify variable importance.  I think an obvious option is like a "Type-3 ANOVA" where you take a full model and extract PDE, then eliminate a covariate, refit and extract PDE, and loop across all variables to get their "Type-3" importance.

**Reply by KGallagher7** [reply]
Thanks for that insight, James! I'll definitely try that out!
